File: backend/miniapp/staff/card_sale.py
"""员工小程序 · 会员购卡"""

from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session
from pydantic import BaseModel
from typing import Optional
from datetime import date

from backend.database import get_db
from backend.models.models import MembershipCard, Member, CardProduct, Sale
from backend.miniapp.common import success, MiniAppException
from backend.miniapp.auth import verify_staff_token
from backend.services.id_gen import generate_id

router = APIRouter(prefix="/cards", tags=["员工-购卡"])


class SellCardRequest(BaseModel):
    member_id: str
    product_name: str
    card_type: str  # 次卡 / 期限卡 / 储值卡 / 现金卡
    total_classes: Optional[int] = 0
    bonus_classes: Optional[int] = 0
    amount: float
    paid_amount: float
    salesperson: Optional[str] = ""
    validity_days: Optional[int] = 365
    start_date: Optional[str] = ""
    remark: Optional[str] = ""


@router.get("/products")
async def list_card_products(
    db: Session = Depends(get_db),
    staff: dict = Depends(verify_staff_token),
):
    """可售卡种列表。"""
    products = db.query(CardProduct).filter(
        CardProduct.status.in_(["上架", "正常"])
    ).all()
    return success(data=[
        {
            "id": p.id,
            "name": p.card_name,
            "price": float(p.price) if p.price else 0,
            "validity_days": p.duration_days or 0,
        }
        for p in products
    ])


@router.post("/sell")
async def sell_card(
    body: SellCardRequest,
    db: Session = Depends(get_db),
    staff: dict = Depends(verify_staff_token),
):
    """会员购卡。"""
    member = db.query(Member).filter(Member.member_id == body.member_id).first()
    if not member:
        raise MiniAppException(2002, "会员不存在")

    card_id = generate_id("MC", db, MembershipCard.card_id)

    card = MembershipCard(
        card_id=card_id,
        member_id=body.member_id,
        member_name=member.name,
        card_name=body.product_name,
        card_type=body.card_type,
        total_classes=body.total_classes or 0,
        bonus_classes=body.bonus_classes or 0,
        remaining_classes=(body.total_classes or 0) + (body.bonus_classes or 0),
        price=body.amount,
        actual_amount=body.paid_amount,
        staff_name=body.salesperson,
        duration_days=body.validity_days or 365,
        start_date=body.start_date or "",
        remark=body.remark,
        status="正常",
    )
    db.add(card)

    # 次卡售出后累加剩余次数到会员
    if body.card_type == "次卡":
        add_classes = (body.total_classes or 0) + (body.bonus_classes or 0)
        if add_classes:
            member.remaining_lessons = (member.remaining_lessons or 0) + add_classes

    # 同时记录售卡到售课表
    sale_id = generate_id("SA", db, Sale.sale_id)
    sale = Sale(
        sale_id=sale_id,
        sale_date=date.today(),
        member_id=body.member_id,
        member_name=member.name,
        course_name=body.product_name,
        bought_hours=0,
        total_price=body.amount,
        actual_amount=body.paid_amount,
        staff_name=body.salesperson,
    )
    db.add(sale)
    db.commit()

    return success(data={
        "card_id": card_id,
        "member_name": member.name,
        "remaining_lessons": member.remaining_lessons or 0,
        "message": "购卡成功",
    })

File: miniapp-staff/pages/card-sale/card-sale.js
// pages/card-sale/card-sale.js
const api = require('../../utils/api');

Page({
  data: {
    memberId: '',
    memberName: '',
    products: [],
    selectedProduct: null,
    form: {
      product_name: '',
      card_type: '次卡',
      total_classes: 0,
      bonus_classes: 0,
      amount: 0,
      paid_amount: 0,
      salesperson: '',
      validity_days: 365,
      remark: '',
    },
    submitting: false,
  },

  onLoad(query) {
    if (query.member_id) {
      this.setData({ memberId: query.member_id });
    }
    this.loadProducts();
    const staffInfo = wx.getStorageSync('staffInfo');
    if (staffInfo) {
      this.setData({ 'form.salesperson': staffInfo.staff_name });
    }
  },

  async loadProducts() {
    try {
      const products = await api.getCardProducts();
      this.setData({ products });
    } catch (e) {}
  },

  selectProduct(e) {
    const idx = e.currentTarget.dataset.index;
    const product = this.data.products[idx];
    if (!product) return;
    this.setData({
      selectedProduct: product,
      'form.product_name': product.name,
      'form.card_type': product.category,
      'form.total_classes': product.total_classes,
      'form.bonus_classes': product.bonus_classes,
      'form.amount': product.price,
      'form.paid_amount': product.price,
      'form.validity_days': product.validity_days,
    });
  },

  onInput(e) {
    const field = e.currentTarget.dataset.field;
    this.setData({ [`form.${field}`]: e.detail.value });
  },

  async onSubmit() {
    if (!this.data.memberId) {
      wx.showToast({ title: '请先选择会员', icon: 'none' });
      return;
    }
    this.setData({ submitting: true });
    try {
      const result = await api.sellCard({
        member_id: this.data.memberId,
        ...this.data.form,
        total_classes: parseInt(this.data.form.total_classes) || 0,
        bonus_classes: parseInt(this.data.form.bonus_classes) || 0,
        amount: parseFloat(this.data.form.amount) || 0,
        paid_amount: parseFloat(this.data.form.paid_amount) || 0,
        validity_days: parseInt(this.data.form.validity_days) || 365,
      });
      wx.showToast({ title: `购卡成功，剩余${result.remaining_lessons}次` });
      wx.navigateBack();
    } catch (e) {}
    this.setData({ submitting: false });
  },
});
